---
title: 'Biostatistics II: PCA and tSNE'
author: "Achim Tresch, Vlada Milchevskaya"
date: "`r doc_date()`"
output: 
    BiocStyle::html_document:
        toc: true
        highlight: tango
---

```{r lib setup, include=FALSE}
library(ggplot2)
library(plotly)
library("ggfortify")
```

# Exercise 1


```{r twodimentional, eval = FALSE, include = FALSE}
x = rnorm(n = 100, mean = 10, sd = 1)
y = 3*x + 8 + rnorm(n = 100, mean = 0, sd = 1)
plot(x,y)

twodimentional = data.frame(x1 = x, x2 = y)
plot(twodimentional)
rownames(twodimentional) <- paste0("point",  1:100)
save(twodimentional, file = "twodimentional.RData")
twodimentional
```


```{r twodimentional solution, eval = FALSE}
load("../exercises_pca_kmeans/twodimentional.RData")
head(twodimentional)
plot(twodimentional)

pca.obj <- prcomp(twodimentional)
head(pca.obj$x)

ggplot( as.data.frame(pca.obj$x), 
        aes(x=PC1 , y=PC2)) + 
    geom_point(size=2)


pca.obj$rotation # new coordinate in (1,0), (0,1) basis
```

# Exercise 2

```{r, eval = FALSE}
cov(twodimentional)

# explained variance
pca.obj$sdev^2/ sum(pca.obj$sdev^2)

var(as.matrix(twodimentional)  %*% c(1,0))
var(as.matrix(twodimentional)  %*% c(0,1))

# projection onto vector (a,b), example
vec = c(1,3)/sqrt(10)
var(as.matrix(twodimentional)  %*% vec)

pca.obj$rotation

# transformation of the covariance matrix
cov_transformed = t(pca.obj$rotation) %*% cov(twodimentional) %*% pca.obj$rotation
round(cov_transformed, digits = 2)

# compare to the explained variance from above
round(cov_transformed / sum(cov_transformed), digits = 3)
```


# Exercise 3

```{r, eval = FALSE}
load("../exercises_pca_kmeans/data3.RData")
prcomp(x = data3)
prcomp(x = data3, center = FALSE)
```

# Exercise 4

```{r, eval = FALSE}
load("../exercises_pca_kmeans/data4.RData")
prcomp(x = data4)
prcomp(x = data4, scale. = TRUE)
```


# Exercise 5 (Single Cell)

To access count data: colnames are cells, rownames of are genes.
An integer in the `singleCellCounts@raw.data` table represents the number of molecules of each gene in each cell.

```{r}
library(Seurat)
library(Rtsne)
library(dplyr)

load("./singleCellCounts.RData")
singleCellCounts@raw.data[1:5,1:5]
```

Possible normalisations:

```{r}
normCounts_1 <- as.matrix(singleCellCounts@raw.data+00.1)/
    singleCellCounts@meta.data$nUMI
normCounts_2 <- log(as.matrix(singleCellCounts@raw.data+00.1))
normCounts_3 <- log(normCounts_1)
```

PCA
```{r}
res.pca <- prcomp(t(normCounts_3))
plot(res.pca) # elbow plot

data_to_plot = data.frame(PC1 = res.pca$x[,1], PC2 = res.pca$x[,2])
pca.plot <- ggplot(data_to_plot, aes(PC1, PC2)) + 
    geom_point(alpha = 1) + 
    theme_classic() 
pca.plot
```

tSNE

```{r}
tsne.res <- Rtsne(res.pca$x[,1:10], pca = FALSE, perplexity=20, max_iter=1000)
data.frame(Dim1=tsne.res$Y[,1], 
           Dim2=tsne.res$Y[,2], 
           Cluster = singleCellCounts@meta.data[,]$res.0.1) %>%
    ggplot(aes(Dim1, Dim2, color = Cluster)) + geom_point() + 
    theme_classic()
```

Marker gene expression

```{r}
marker_gene_expression = normCounts_3["Neurod6", rownames(res.pca$x)]

data.frame(Dim1=tsne.res$Y[,1], 
           Dim2=tsne.res$Y[,2], 
           Cluster = singleCellCounts@meta.data[,]$res.0.1,
           marker_gene_expression = marker_gene_expression) %>%
    ggplot(aes(Dim1, Dim2, color = marker_gene_expression)) + geom_point() + 
    scale_color_gradient(low = "darkblue", high = "red") + 
    theme_classic()


```















```{r, eval = FALSE, include = FALSE}
normalized_counts <- t(t(log(singleCell$counts + 1)))
normalized_counts[1:5,1:5]; 4/libSize[1]
normalized_counts[1:5,1:5]

pr.nor <- prcomp(t(normalized_counts))
autoplot(pr.nor)
ggplot( as.data.frame(pr.nor$x), aes(x=PC1 , y=PC2, color=libSize)) + 
    geom_point(size=2, alpha = 0.5) + 
    scale_color_gradient(low = "blue", high = "red")

ggplot( as.data.frame(pr.nor$x), 
        aes(x=PC1 , y=PC2, color=singleCell$class_labels)) + 
    geom_point(size=2, alpha = 0.5)
```




```{r data3, eval = FALSE, include = FALSE}
x1 = sample(size = 20, x = c(1:10),replace = TRUE )
data3 <- data.frame( x1 = x1, 
                  x2 = sample(size = 20, x = c(1001:1010),replace = TRUE ))
rownames(data3) <- paste0("point",  1:20)
save(data3, file = "data3.RData")
```


```{r data4, include = FALSE, eval = FALSE}
poit_example <- c(1,1)

data4 <- data.frame(
    x1 = c(1,0,0,0),
    x2 = c(0,1,0,0),
    x3 = c(0,0,1,0),
    x4 = c(0,0,0,100)
)

plot(data4)
prcomp(x = data4)
prcomp(x = data4, scale. = TRUE)
save(data4, file = "data4.RData")
```

